lib/can/src/CANInterface.cpp

```diff
@@ -0,0 +1,157 @@
+#include "CANInterface.h"
+
+CANInterface::CANInterface(const Config &config)
+    : m_CANBus1(config.can1_RX, config.can1_TX, config.frequency_hz),
+      m_CANBus2(config.can2_RX, config.can2_TX, config.frequency_hz),
+      m_activeCANBus(&m_CANBus1),
+      m_rxPostmanThread(RX_POSTMAN_THREAD_PRIORITY),
+      m_rxClientThread(RX_CLIENT_THREAD_PRIORITY),
+      m_txProcessorThread(TX_PROCESSOR_THREAD_PRIORITY),
+      m_rxStreamedMsgMap(config.rxStreamedMsgMap),
+      m_txStreamedMsgMap(config.txStreamedMsgMap),
+      m_rxOneShotMsgHandler(config.rxOneShotMsgHandler) {
+  // Set hw filters
+  m_CANBus1.setFilter(config.filterID, config.filterFormat, config.filterMask);
+  m_CANBus2.setFilter(config.filterID, config.filterFormat, config.filterMask);
+
+  // Put CAN bus 2 in silent monitoring mode
+  m_CANBus2.monitor(true);
+
+  // Processing threads
+  m_rxPostmanThread.start(callback(&m_rxEventQueue, &EventQueue::dispatch_forever));
+  m_rxClientThread.start(callback(this, &CANInterface::rxClient));
+  m_txProcessorThread.start(callback(this, &CANInterface::txProcessor));
+
+  // RX IRQ
+  m_CANBus1.attach(callback(this, &CANInterface::rxISR), CAN::RxIrq);
+  m_CANBus2.attach(callback(this, &CANInterface::rxISR), CAN::RxIrq);
+}
+
+void CANInterface::rxISR(void) {
+  can_irq_set(m_activeCANBus->getHandle(), IRQ_RX, false);
+  m_rxEventQueue.call(callback(this, &CANInterface::rxPostman));
+}
+
+void CANInterface::rxPostman(void) {
+  CANMsg msg;
+  // this loop is needed to avoid missing msg received between turning off the IRQ and turning it back on
+  while (m_activeCANBus->read(msg)) {
+    CANMsg *mail = m_rxMailbox.try_alloc();  // no wait
+
+    if (mail != nullptr) {
+      *mail = msg;
+      MBED_ASSERT(m_rxMailbox.put(mail) == osOK);
+    } else {
+      MBED_WARNING(MBED_MAKE_ERROR(MBED_MODULE_PLATFORM, MBED_ERROR_CODE_BUFFER_FULL), "CAN RX mailbox full");
+    }
+  }
+  can_irq_set(m_activeCANBus->getHandle(), IRQ_RX, true);
+}
+
+void CANInterface::rxClient(void) {
+  while (true) {
+    CANMsg *mail = nullptr;
+
+    do {
+      mail = m_rxMailbox.try_get();  // TODO: try_get_for was not working. Investigate why and use it
+      ThisThread::sleep_for(1ms);
+    } while (mail == nullptr);
+
+    // Extract message
+    CANMsg msg = *mail;
+    MBED_ASSERT(m_rxMailbox.free(mail) == osOK);
+
+    // If message is one-shot, process message
+    if (m_rxOneShotMsgHandler->find(msg.getID()) != m_rxOneShotMsgHandler->end()) {
+      m_rxOneShotMsgHandler->at(msg.getID())(msg);
+    }
+
+    // Otherwise message is streamed - extract message signals and put in RX message map
+    else {
+      HWBRIDGE::CANMsgData_t msgData;
+      msg.getPayload(msgData);
+
+      if (!HWBRIDGE::unpackCANMsg(msgData.raw, msg.getID(), m_rxStreamedMsgMap)) {
+        MBED_WARNING(MBED_MAKE_ERROR(MBED_MODULE_PLATFORM, MBED_ERROR_CODE_INVALID_DATA_DETECTED),
+                     "CAN RX message unpacking failed");
```

Comment:
I made these all MBED_WARNINGs instead of MBED_ERRORs because message packing and mailbox full faults shouldn't crash the program